Automattic/jetpack
# Lazy Images: Move into a package

#### Changes proposed in this Pull Request:
* Moves the Lazy Images module into a package. @gravityrail should be able to provide more context. It has become a requirement for an (yet) undisclosed A8C project.

#### Does this pull request change what data or activity we track or use?
No it should not.

#### Testing instructions:






* Go to `wp-admin/admin.php?page=jetpack_modules`
* Activate the "Lazy images" module
* Create a new post with an image inside it.
* Open that post as a non logged in user
* Look at the source of the page and verify that the img tag as the `data-lazy-loaded="1"` property.

#### Proposed changelog entry for your changes:
* Moves the Lazy Images module into a package.

#### Additional development context:

Here are some additional information related to the decision I have made for migrating the `Lazy Images` module into a standalone package. I have made some assumptions which might be incorrect and therefore some changes will need to be amended based on your guidance.

- While migrating the module to a package, I tried to remove or update references to the module source file in order place of the repository such as from:
    - `.eslintignore`
    - `modules/photon-cdn/jetpack-manifest.php`: I am not sure if this one would be generated automatically so i just updated the path to the JavaScript files manually.
   - `phpunit.xml.dist`
   - `modules/lazy-images.php`: I kept this file even though the module source moved into a package as I wanted to make sure the module pattern activation/deactivation still work. I have also moved some of the credit comments into the package itself (`packages/lazy-images/src/lazy-images.php`) as it seems appropriate to belong there as the package can be a standalone download in any project.

- In `packages/lazy-images/src/lazy-images.php`:
    - I have used the version `8.8` in the PHPDoc `@since` entry but I am unsure if this would be the correct version.
    - I have created a new method  called `is_amp_request` because as a standalone package it cannot rely anymore on the `Jetpack_AMP_Support::is_amp_request()` method.
    - I have created a new method  called `get_file_url_for_environment` because as a standalone package it cannot rely anymore on the `Assets::get_file_url_for_environment` method as the JS asset is located within the package.
    - I have fixed the coding standards.

- In `packages/lazy-images/tests/php/test_class.lazy-images.php`:
    - I have fixed the coding standards.

- In `packages/lazy-images/src/js/lazy-images.js`:
    - I have fixed a spelling mistake
    - I have fixed Fix an issue with a comma being in place of a semi-column ( see commit 746e75e7b)

- I have introduced some tooling for building assets, running unit tests and linters. The reason is because I was not sure if/how to integrate these within the Jetpack workflow. So these changes might need to be altered.

    - Previously the minified version of the `lazy-images.js` file was generated by the Jetpack build workflow. Here I have introduced a Webpack config to build the minified version. The `yarn build` command allows building the minified version of the JavaScript.
    - Because the PHP test class (`packages/lazy-images/tests/php/test_class.lazy-images.php`) extends the `WP_UnitTestCase` class, it seems that the `wordpress-tests-lib` is required. Therefore I also needed to have a bootstrapped  WordPress testing environment which I am creating with Docker. The `yarn test:phpunit` command allows to run the PHP unit tests.
    - I have introduced linting the PHP code. The `yarn php:lint` command allows to run the lint the PHP code base.

As I mention some of my assumptions and decisions above might be incorrect and will need to be adapted to an existing Jetpack workflow.
